### PR TITLE
WIP: Project 3 Dynamics Report

### PR DESCRIPTION
**Pull request by lynnzhao92**
All the members are contributing to this pull request and working on the reports together. @harryh1993 @Chenyu-Renee @HongfeiZhao @wyshi @mingyungkim Please constantly review and edit on the report.

### CODE REVIEW

File: doc/source/report/dynamics.rst [hunk 1]
```diff
@@ -3,73 +3,147 @@
...
 
--  We need labels of states of interest with respect to the time intervals in a specific mouse day to extract the structure of the Markov Chain.
+- Divide each mouse day into small time intervals, say 5 minutes. (The time interval is optimized in the future analysis)
+- For each of the small time intervals, aggregate the data from all mouses in the same strain for all mouse days and estimate the transition probability matrix of a discrete homogeneous Markov Chain model just for this small time interval.
+- each of these transition probability matrices is estimated by MLE method, where e.g.: 
+   .. math:: P(F_{t+1} | W_{t}) = \frac{N_{WF}}{N_{W.}}
+   where ``N_{WF}`` indicates the counts of transitions from W to F and ``N_{W.}`` indicates the counts of transitions starting from W, no matter where it ends.
```
Comment (jarrodmillman) [inline doc/source/report/dynamics.rst]: You should use

```
$N_{WF}$
```

Sphinx will render it for the website and PDF.

**Comment by mingyungkim** [comment]
@jarrodmillman Hello Jarrod, this pull request is ready to merge. The group members have reviewed it. Please review and merge it. Thanks!

File: doc/source/report/dynamics.rst [hunk 2]
```diff
@@ -3,60 +3,149 @@
 Dynamics of AS Patterns
 =======================
 
-Statement of Problem
---------------------
+Introduction
+------------
 
-The objective of Dynamics Analysis is to analyze and characterize the behaviors of different strains of mice using Markov Chain model. We could use the MC models to further explore the discrepancy between different strains and even simulate a particular mouse day for a particular strain.
+The objective of Dynamics Analysis is to analyze and characterize the state transforming behaviors of different strains of mice. The analysis is mainly focused on the three strains of mice included in the interval dataset. Using Markov Chain Model, a sequence of state transition probability matrix is retrieved for each strain depends on the time intervals and a simulation mouse is generated based on the model. A evaluation system is created for comparing the behaviors of two certain mice, and based on which the optimal time interval is choosen so as to have the most realistic simulation mice.Further visualizatoin on those fake mice is done to explore the discrepancy between different strains's behaviors.
```
Comment (alannaiverson) [inline doc/source/report/dynamics.rst]: Hey guys! Jarrod asked me to help review the wording of some of the reports so I'll comment with some suggestions following this one. Feel free to take them or leave them and I hope you don't see this as imposing! 
-Alanna

"The objective of Dynamics Analysis is to analyze and characterize the state transforming behaviors of different strains of mice. The analysis is mainly focused on the three strains of mice included in the interval dataset. Using a Markov Chain Model, a sequence of state transition probability matrix is retrieved for each strain depending on the time intervals. A simulation mouse is then generated based on the model and an evaluated. The evaluation system is created using the time interval that produces the most realistic simulation mouse. Further visualization of the fake mice is done to explore the discrepancy between different strain's behaviors."

Comment (mingyungkim) [inline doc/source/report/dynamics.rst]: @alannaiverson Thanks Alanna. We will revise our report following your suggestions.

File: doc/source/report/dynamics.rst [hunk 3]
```diff
@@ -3,60 +3,149 @@
...
 
-The statistical problems in this project are mainly naturally divided into two parts, the modeling part and the simluation of the model part. The modelling part consists of estimating the probability of events transition using the raw data of mouse's spatial and temporal data, and once we've got the model, we could use this model to do predictions and simulate a fake mouse day.
+- Divide each mouse day into small time intervals, say 5 minutes. (The time interval is optimized in the future analysis)
+- For each of the small time intervals, aggregate the data from all mouses in the same strain for all mouse days and estimate the transition probability matrix of a discrete homogeneous Markov Chain model just for this small time interval.
```
Comment (alannaiverson) [inline doc/source/report/dynamics.rst]: "mouses" should be "mice"

File: doc/source/report/dynamics.rst [hunk 4]
```diff
@@ -3,60 +3,149 @@
...
------------------------------
 
--  We need labels of states of interest with respect to the time intervals in a specific mouse day to extract the structure of the Markov Chain.
 
-Methodology/ Approach Description
----------------------------------
+4. Simluation: 
 
-1. Defining states of interest: which behaviors of the mouse are we interested in? 
+Use the Transition Model generated to simluate a typical mouse day for a typical strain. The simulation is time sensitive which means the simulation mouse is generated depend on the time interval chosen in the transition model. The time interval length is optimized in the future analysis so as the generate the best fake mouse for each strain.
```
Comment (alannaiverson) [inline doc/source/report/dynamics.rst]: "depend" should be "depending"

File: doc/source/report/dynamics.rst [hunk 5]
```diff
@@ -3,60 +3,149 @@
...
 
-   -  Data cleaning: Clean the raw intervals given by the measurements in the experiment into interval data that makes more sense and consistent. Also need to check if any of our states overlapped.
-   -  Data reformating: Convert the cleaned interval data into strings of events or matrices containing both information from timestamps and the events at those timepoints.
+5. Evaluation System:
 
-3. Estimating Transition Probability: Estimate the transition probability matrix of the Markov Chain using the data given. One of the key challenges to estimate the transition matrix is that our model is actually time continuous non-homogeneous Markov Chain, and the parameters are too difficult to estimate given the data we have. Instead, we figured out a way to make our model a composite of small homogeneous discrete time Markov Chains, so that we could perform a rough estimation of the original time continuous non-homogeneous Markov Chain. We followed the following steps to achieve our goal:
+How to get a score for the performance of our simulation?
+The evaluation system is trying to create a score to evaluate the performance of the simulation. Higher the score, better
```
Comment (alannaiverson) [inline doc/source/report/dynamics.rst]: "The higher the score, the better..."

Sorry that all of these are such small corrections, I know it's annoying!

File: doc/source/report/dynamics.rst [hunk 6]
```diff
@@ -3,60 +3,149 @@
...
+the simulation is doing in catching the pattern in the mouse day. Because we have unbalanced data, meaning most of the
+status in one typical mouse day are IS and OTHERS and only a few of them are DRINK and EAT, we should give more weights on
+correct predictions about DRINK and EAT. We first calculate the proportion of the four status within different days we have
```
Comment (alannaiverson) [inline doc/source/report/dynamics.rst]: "status" should be "statuses"

File: doc/source/report/dynamics.rst [hunk 7]
```diff
@@ -3,60 +3,149 @@
...
+ranges from 0 to 1 for our data.
 
-Remarks:
---------
-Testing for each function should be done by the person who wrote the function.
 
-References:
+6. Comparison:
+
+The comparison function plots the simulated behavior dynamics. It plots the the simulated behavior dynamics obtained from the evaluation function for the given strain during the given time period.
```
Comment (alannaiverson) [inline doc/source/report/dynamics.rst]: "The comparison function plots the simulated behavior dynamics obtained from the evaluation function for the given strain during the given time period."

File: doc/source/report/dynamics.rst [hunk 8]
```diff
@@ -3,60 +3,149 @@
...
+
+The x-axis is the time range, and the default time range is from 40,000 to 40,100. It is because the behavior dynamics show clear differences across the strains during this time range and because the behavior dynamics have more active actions than the other time period. Users can, of course, define the time range of their interest. But they should be aware of the lower and upper bounds of the time range before doing comparisons. For the first strain (strain_num = 0), the time range should be between 1 and 92,400. For the second one (strain_num = 1), it should be between 1 and 90,000. The time range should be between 1 and 88,800 for the last strain (strain_num = 2). 
+
+Of note is that, based on our understadning and explanations from the research team, 0 indicates the time when a mouse wakes up and starts its day. The unit of time is understood as one second. Thus, the default time range (from 40,000 to 40,100) represents around 11 hours after the start of the day. Based on this defnintion of the time range, users can change it and compare simulated behavior dynamics across the three strains.
```
Comment (alannaiverson) [inline doc/source/report/dynamics.rst]: "understadning" = "understanding"
"defnintion" = "definition"

File: doc/source/report/dynamics.rst [hunk 9]
```diff
@@ -3,60 +3,149 @@
...
+
+We do not define the y-axis for this comparison function, because our interest is to visually understand the dynamics of the four states during the given time period. For better understanding of the behavior dynamics, we assign different colors for different states. In the plot, blue represents IS, skyblue represents eating, yellow represents drinking, and red represents other activities in AS.
+
+
+Result
+------
+
+The problem we are insteresting in here is whether the three strains of mice are indeed acting differently in a time series manner. The behaviors are compared using the three simulation mice, each for one strain. Therefore, as the first step, the optimal time interval is selected so as to have the most-real simulation mouse that behaves the most similarly to its strain, which is evaluated using the score system we created. As a result, the best time interval length is selected for each strain as well as the corresponding simulation mice behavior and the comparison score.
```
Comment (alannaiverson) [inline doc/source/report/dynamics.rst]: "interesting" = "interested"

File: doc/source/report/dynamics.rst [hunk 10]
```diff
@@ -3,60 +3,151 @@
...
+
+    find_best_interval(data_df,strain=0)
+    # (600, array([0, 0, 0, ..., 0, 0, 0]), 0.70509736459572225)
+    find_best_interval(data_df,strain=1)
+    # (600, array([0, 0, 0, ..., 0, 0, 0]), 0.91423472578532516)
+    find_best_interval(data_df,strain=2)
+    # (1800, array([0, 0, 0, ..., 0, 0, 0]), 0.83975976073161329)
+
+
+As the scipt suggested, the best time interval selected is 600 seconds for the first strain which generates the simultation mouse that has 71% similarity compared to the real mice in that strain, and the simulation behavior is quoted in the numpy array format. Similarly, it shows the 600s optimal time interval with 91% similarity for the second strain and the 1800s optimal time interval with 84% similarity for the third strain.
```
Comment (alannaiverson) [inline doc/source/report/dynamics.rst]: "As the script suggested, the best time interval selected is 600 seconds for the first strain. This generates a simulation mouse that is 71% similar to the real mice in that strain. Note that the simulation behavior is quoted in the numpy array format. Similarly, it shows the 600s optimal time interval with 91% similarity for the second strain and the 1800s optimal time interval with 84% similarity for the third strain."

File: doc/source/report/dynamics.rst [hunk 11]
```diff
@@ -3,60 +3,151 @@
...
+    find_best_interval(data_df,strain=2)
+    # (1800, array([0, 0, 0, ..., 0, 0, 0]), 0.83975976073161329)
+
+
+As the scipt suggested, the best time interval selected is 600 seconds for the first strain which generates the simultation mouse that has 71% similarity compared to the real mice in that strain, and the simulation behavior is quoted in the numpy array format. Similarly, it shows the 600s optimal time interval with 91% similarity for the second strain and the 1800s optimal time interval with 84% similarity for the third strain. 
+
+With the resulting best simulation mice, we move forward to compare the between-strain behaviors using visulization. The simulated mice show different behavior dynamics across the strains, during the given time period (36,000 to 36,100; around 10 hours after the start of the day). The fake mouse from the first strain does the other active actions and stays inactive for a short period of time. The fake mice from the other two strains, on the other hand, show different dynamic patterns: do the other active actions, consume food, and stay inactive. The fake mice from the third strain 
+
```
Comment (alannaiverson) [inline doc/source/report/dynamics.rst]: I think the sentence starts with "The fake mice from the third strain..." is missing its ending?

File: doc/source/report/dynamics.rst [hunk 12]
```diff
@@ -65,11 +156,12 @@ https://github.com/hmmlearn/hmmlearn
...
+------------
 
 - Data Preprocessing: Hongfei
 - Modeling: Jianglong
 - Simlation: Chenyu
-- Evaluation of length of small time interval chosen: Weiyan, Lynn, Mingyung
-- Final Report writeup: Weiyan, Lynn, Mingyung
+- Score: Weiyan
+- Evaluation: Luyun Zhao
+- Comparison: Mingyung
```
Comment (alannaiverson) [inline doc/source/report/dynamics.rst]: Looks great guys!

File: doc/source/report/dynamics.rst (hunk 4 above)
Comment (mingyungkim) [inline doc/source/report/dynamics.rst]: Thank you:)

File: doc/source/report/dynamics.rst (hunk 5 above)
Comment (mingyungkim) [inline doc/source/report/dynamics.rst]: Yes, I changed that. Thank you:)

File: doc/source/report/dynamics.rst (hunk 6 above)
Comment (mingyungkim) [inline doc/source/report/dynamics.rst]: Thank you:)

File: doc/source/report/dynamics.rst (hunk 7 above)
Comment (mingyungkim) [inline doc/source/report/dynamics.rst]: I followed your expression. Thanks.

File: doc/source/report/dynamics.rst (hunk 8 above)
Comment (mingyungkim) [inline doc/source/report/dynamics.rst]: Thank you for finding typos.

File: doc/source/report/dynamics.rst (hunk 9 above)
Comment (mingyungkim) [inline doc/source/report/dynamics.rst]: Thank you for checking grammars.

File: doc/source/report/dynamics.rst (hunk 10 above)
Comment (mingyungkim) [inline doc/source/report/dynamics.rst]: I followed your expression. Thanks:)

File: doc/source/report/dynamics.rst (hunk 3 above)
Comment (mingyungkim) [inline doc/source/report/dynamics.rst]: Yes. I changed it.

**Comment by mingyungkim** [comment]
We have reviewed the report based on Alanna's comments, and this pull request is ready to be merged if the conflicts are resolved.

**Comment by mingyungkim** [comment]
It has been rebased at #196.

**Pull request closed without merging**